File: image_utils-win32/image_utils.c
#include <stdlib.h>
#include <string.h>
#include <FreeImage.h>
#include "ruby.h"

static int ico2png(char **output, size_t *output_length, const char *input, size_t input_length, int width, int height)
{
	FIMEMORY *mem_src = NULL;
	FREE_IMAGE_FORMAT fif = 0;
	int flags = 0;
	FIBITMAP *icon = NULL;
	FIBITMAP *resized = NULL;
	FIMEMORY *mem_dest = NULL;
	char *mem_dest_ptr = NULL;

	mem_src = FreeImage_OpenMemory((BYTE *)input, (DWORD)input_length);
	fif = FreeImage_GetFileTypeFromMemory(mem_src, 0);
	if (fif == FIF_UNKNOWN) {
		fif = FIF_ICO;
	}
	switch (fif) {
	case FIF_ICO:
		flags = ICO_MAKEALPHA;
		break;
	case FIF_JPEG:
		flags = JPEG_ACCURATE;
		break;
	}
	icon = FreeImage_LoadFromMemory(fif, mem_src, flags);
	FreeImage_CloseMemory(mem_src);
	if (icon == NULL) {
		return 0;
	}
	resized = FreeImage_Rescale(icon, width, height, FILTER_CATMULLROM);
	FreeImage_Unload(icon);
	if (resized == NULL) {
		return 0;
	}
	mem_dest = FreeImage_OpenMemory(NULL, 0);
	if (!FreeImage_SaveToMemory(FIF_PNG, resized, mem_dest, 0)) {
		FreeImage_CloseMemory(mem_dest);
		FreeImage_Unload(resized);
		return 0;
	}
	FreeImage_Unload(resized);
	if (!FreeImage_AcquireMemory(mem_dest, (BYTE **)&mem_dest_ptr, (DWORD *)output_length)) {
		FreeImage_CloseMemory(mem_dest);
		return 0;
	}
	if ((*output = (char *)malloc(*output_length)) == NULL) {
		FreeImage_CloseMemory(mem_dest);
		return 0;
	}
	memcpy((void *)*output, (void *)mem_dest_ptr, *output_length);
	FreeImage_CloseMemory(mem_dest);
	return 1;
}

static VALUE ruby_ico2png(VALUE self, VALUE ico)
{
	char *png;
	size_t length;
	VALUE result;

	StringValue(ico);
	if (RSTRING(ico)->ptr == NULL || RSTRING(ico)->len == 0) {
		return Qnil;
	}
	if (!ico2png(&png, &length, RSTRING(ico)->ptr, RSTRING(ico)->len, 16, 16)) {
		return Qnil;
	}
	result = rb_str_new(png, length);
	free(png);
	return result;
}

void Init_image_utils(void)
{
	VALUE module = rb_define_module("ImageUtils");
	rb_define_method(module, "ico2png", ruby_ico2png, 1);
}

File: fastladder/lib/image_utils.rb
require "dl/import"
require "dl/struct"

module ImageUtils
  module FreeImage
    extend DL::Importable
    FIBITMAP = struct ["void* data"]
    FIMEMORY = struct ["void* data"]
    typealias "FREE_IMAGE_FORMAT", "int"
    typealias "FREE_IMAGE_FILTER", "int"
    begin
      dlload "libfreeimage.so"
    rescue => ex
      dlload "lib/libfreeimage-3.10.0.dylib"
    end
    extern "FIMEMORY* FreeImage_OpenMemory(BYTE*, DWORD)"
    extern "void FreeImage_CloseMemory(FIMEMORY*)"
    extern "BOOL FreeImage_AcquireMemory(FIMEMORY*, BYTE**, DWORD*)"
    extern "FIBITMAP* FreeImage_LoadFromMemory(FREE_IMAGE_FORMAT, FIMEMORY*, int)"
    extern "BOOL FreeImage_SaveToMemory(FREE_IMAGE_FORMAT, FIBITMAP*, FIMEMORY*, int)"
    extern "FREE_IMAGE_FORMAT FreeImage_GetFileTypeFromMemory(FIMEMORY*, int)"
    extern "FIBITMAP* FreeImage_Rescale(FIBITMAP*, int, int, FREE_IMAGE_FILTER)"
    extern "void FreeImage_Unload(FIBITMAP*)"
    FIF_UNKNOWN = -1
    FIF_BMP = 0
    FIF_ICO = 1
    FIF_JPEG = 2
    FIF_PNG = 13
    FIF_GIF = 25
    ICO_MAKEALPHA = 1
    JPEG_ACCURATE = 2
    FILTER_CATMULLROM = 4
  end
  module_function
  
  FILE_FORMATS = {
    "bmp" => FreeImage::FIF_BMP,
    "icon" => FreeImage::FIF_ICO,
    "ico" => FreeImage::FIF_ICO,
    "jpeg" => FreeImage::FIF_JPEG,
    "jpg" => FreeImage::FIF_JPEG,
    "png" => FreeImage::FIF_PNG,
    "gif" => FreeImage::FIF_GIF,
  }
  def file_format(str)
    FILE_FORMATS[str.to_s.downcase]
  end

  def convert(src, options = {})
    src_format = options[:src_format] ? file_format(options[:src_format]) : nil
    unless dest_format = (options[:dest_format] ? file_format(options[:dest_format]) : src_format)
      return nil
    end
    width = options[:width] ? options[:width].to_i : nil
    height = options[:height] ? options[:height].to_i : nil

    unless mem_src = FreeImage::freeImage_OpenMemory(src, src.length)
      return nil
    end
    fif = FreeImage::freeImage_GetFileTypeFromMemory(mem_src, 0);
    if fif == FreeImage::FIF_UNKNOWN
      return nil if src_format.nil?
      fif = src_format
    end
    flags = 0
    case (fif)
    when FreeImage::FIF_ICO
      flags |= FreeImage::ICO_MAKEALPHA
    when FreeImage::FIF_JPEG
      flags |= FreeImage::JPEG_ACCURATE
    end
    bmp = FreeImage::freeImage_LoadFromMemory(fif, mem_src, flags)
    FreeImage::freeImage_CloseMemory(mem_src)
    if bmp.nil?
      return nil
    end
    if width and height
      resized_bmp = FreeImage::freeImage_Rescale(bmp, width, height, FreeImage::FILTER_CATMULLROM);
      FreeImage::freeImage_Unload(bmp)
      unless bmp = resized_bmp
        return nil
      end
    end
    mem_dest = FreeImage::freeImage_OpenMemory(nil, 0)
    unless FreeImage::freeImage_SaveToMemory(dest_format, bmp, mem_dest, 0)
      FreeImage::freeImage_CloseMemory(mem_dest)
      FreeImage::freeImage_Unload(bmp)
      return nil
    end
    FreeImage::freeImage_Unload(bmp)
    mem_dest_ptr = DL::malloc(DL::sizeof("p"))
    mem_dest_len = DL::malloc(DL::sizeof("i"))
    unless FreeImage::freeImage_AcquireMemory(mem_dest, mem_dest_ptr, mem_dest_len)
      FreeImage::freeImage_CloseMemory(mem_dest)
      return nil
    end
    length = mem_dest_len.to_a("I", 1)[0]
    dest = DL::PtrData::new(mem_dest_ptr.to_a("P", 1)[0]).to_s(length)
    #GC.start
    dest
  end

  def self.ico2png(src)
    convert(src, :src_format => :ico, :dest_format => :png, :width => 16, :height => 16)
  end
end

if __FILE__ == $0
  require "test/unit"
  class ImageUtilsTest < Test::Unit::TestCase # :nodoc:
    def test_ico2png
      png = ImageUtils::ico2png(File.read("favicon.ico"))
      assert_equal(png, File.read("favicon.png"))
    end
  end
end
